### PR TITLE
Use guard in recursion functions

### PR DESCRIPTION
**Pull request by AMHOL**

### CODE REVIEW

File: lib/transproc/recursion.rb [hunk 1]
```diff
@@ -30,13 +32,10 @@ module Recursion
     # @api public
     def array_recursion(value, fn)
       result = fn[value]
+      guarded = Transproc(:guard, -> v { v.is_a?(::Array) }, -> v { Transproc(:array_recursion, fn)[v] })
```
Comment (solnic) [inline lib/transproc/recursion.rb]: this should be memoized otherwise every time you call it guard function gets created, same applies to hash_recursion

Comment (AMHOL) [inline lib/transproc/recursion.rb]: Well spotted, will have a look tonight

Comment (solnic) [inline lib/transproc/recursion.rb]: We could maybe introduce type-check functions. `Transproc.register(:is_array, -> v { v.is_a?(Array) })`?? /cc @splattael

Comment (splattael) [inline lib/transproc/recursion.rb]: Interesting! Would `Transproc.register(:array?, -> v { v.is_a?(Array) })` work?

Comment (solnic) [inline lib/transproc/recursion.rb]: Not OOTB, it requires a bit more work. I'll experiment with that.

**Pull request merged**